modules/imgproc/src/drawing.cpp

```diff
@@ -1187,18 +1182,18 @@ FillConvexPoly( Mat& img, const Point2l* v, int npts, const void* color, int lin
 
                     for(;;)
                     {
-                        ty = (int)((v[idx].y + delta) >> shift);
+                        ty = (int)((v[idx].y + XY_HALF) >> XY_SHIFT);
                         if( ty > y || edges == 0 )
+                        {
+                            xe = v[idx].x;
                             break;
+                        }
                         xs = v[idx].x;
                         idx += di;
-                        idx -= ((idx < npts) - 1) & npts;   /* idx -= idx >= npts ? npts : 0 */
+                        if (idx >= npts) idx -= npts;
                         edges--;
                     }
 
-                    xs <<= XY_SHIFT - shift;
-                    xe = v[idx].x << (XY_SHIFT - shift);
```

Comment:
Probably vc10 compiler bug is somewhere here. Adding "printf" statement repairs regression test.